pypeit/core/flat.py

```diff
@@ -19,8 +19,353 @@
+    fullbkpt = bsp_set.breakpoints
+    spat_set, outmask_spat, spatfit, _ = utils.bspline_profile(ximg_in, normin, np.ones_like(normin),np.ones_like(normin),
+                                                               nord=4,upper=5.0, lower=5.0,fullbkpt = fullbkpt)
+
+    if debug:
+        plt.clf()
+        ax = plt.gca()
+        ax.plot(ximg_fit, norm_spec_fit, color='k', marker='o', markersize=0.4, mfc='k', fillstyle='full',linestyle='None',
+                label = 'all pixels')
+        if imed is not None: # If we computed a median show the pixels we used
+            ax.plot(ximg_fit[~imed], norm_spec_fit[~imed], color='red', marker='+',markersize=1.5, mfc='red',
+            fillstyle='full', linestyle='None', label = 'masked')
+            if no_illum is True:
+                label =  'median spatial profile, NOT USED'
+            else:
+                label = 'median spatial profile'
+            ax.plot(ximg_fit[imed], normimg, color='lawngreen', label = label)
+        ax.plot(ximg_in, spatfit, color='cornflowerblue', label = 'final slit illumination function')
+        ax.set_ylim((np.fmax(0.8 * spatfit.min(), 0.5), 1.2 * spatfit.max()))
+        ax.set_xlim(-0.02, 1.02)
+        plt.legend()
+        plt.xlabel('Normalized Slit Position')
+        plt.ylabel('Normflat Spatial Profile')
+        plt.show()
+
+    # Evaluate and save
+    illumflat = np.ones_like(flat)
+    illumflat[thismask], _ = spat_set.value(ximg[thismask])
+    norm_spec_spat = np.ones_like(flat)
+    norm_spec_spat[thismask] = flat[thismask]/np.fmax(spec_model[thismask], 1.0)/np.fmax(illumflat[thismask],0.01)
+    msgs.info('Performing illumination + scattered light flat field fit')
+
+    # Flat field pixels for fitting spectral direction
+    isrt_spec = np.argsort(piximg[thismask])
+    pix_twod = piximg[thismask][isrt_spec]
+    ximg_twod = ximg[thismask][isrt_spec]
+    norm_twod = norm_spec_spat[thismask][isrt_spec]
+
+    fitmask = inmask[thismask][isrt_spec] & (np.abs(norm_twod - 1.0) < 0.30)
+    # Here we ignore the formal photon counting errors and simply assume that a typical error per pixel.
+    # This guess is somewhat aribtrary. We then set the rejection threshold with sigrej_illum
+    var_value = 0.01
+    norm_twod_ivar = fitmask.astype(float)/(var_value**2)
+    sigrej_illum = 4.0
+
+    poly_basis = pydl.fpoly(2.0*ximg_twod - 1.0, npoly).T
+
+    # Perform the fill 2d fit now
+    twod_set, outmask_twod, twodfit, _ = utils.bspline_profile(pix_twod, norm_twod, norm_twod_ivar,poly_basis,
+                                                               inmask = fitmask, nord = 4,
+                                                               upper=sigrej_illum, lower=sigrej_illum,
+                                                               kwargs_bspline = {'bkspace':spec_samp_coarse},
+                                                               kwargs_reject={'groupbadpix':True, 'maxrej': 10})
+
+    if debug:
```

Comment:
Might be nice to hide these debug bits
as separate methods.

Reply:
Yes, these are placeholders for QA output on the flat fielding, which is not coded up yet.